### ISSUE TITLE
Incremental tree-style hinting

### ISSUE THREAD

**Issue by hendursaga** [atlas-engineer/nyxt#2509]
In Emacs, there's a couple of packages such as [avy](https://github.com/abo-abo/avy) where you can search for text "incrementally" in a character-by-character decision tree. I would love a similar feature for the hinting system (in addition to general search, too). That way, I could more easily select which hints I actually want, and faster. Also, it could help out in the cases where hints are too close together and it's hard to tell what sequence of letters they are.

**Comment by Ambrevar** [comment]
Isn't it already what we do with hints?  How is tree-style hinting different?  Can you give an example?

**Comment by hendursaga** [comment]
The incremental part we don't appear to have, and also this is narrowed to whatever is displayed on screen. Say you had 33 elements on screen and the Latin alphabet for your hints. That would be, what, 8 hints that start with the letter A? So you press A and then now all the other hints either fade or go away entirely. And it doesn't have to be limited to what `avy` uses. I could see you hinting all clickable links, and then from that subset, narrow it down further...

**Comment by aadcg** [comment]
There's an interesting idea here. When the length of the hint overlay string is bigger than one, the hint could be followed incrementally indeed. 

This would only make sense when filtering only takes into account the hint overlays strings. Our default interface, the expanded prompt buffer with the ability to match against several attributes, would not benefit from this as it wouldn't be clear if "A" means "look for the substring A in the object attributes" or "filter through the hint overlay that starts with A".

**Comment by hendursaga** [comment]
> This would only make sense when filtering only takes into account the hint overlays strings.

That much is true, yes. I thought we were working on a much more advanced search & filtering interface, though?? Selecting a field, sorting it, removing it, etc? I'm thinking [VisiData](https://www.visidata.org/) here.. crazy powerful tool! Naturally we wouldn't need statistical tools etc but it's great inspiration!

**Comment by aartaka** [comment]
> This would only make sense when filtering only takes into account the hint overlays strings.

Not necessarily. We can exclude the clearly-not-matching candidates from the suggestions, and make them disappear from the viewport in the default hint interface too. I can imagine searching for "code" and getting only hints that contain the word or the ones that are close enough in the letter combinations.

**Comment by aadcg** [comment]
That is possible. The question is whether it'd be a good default. I think we should be very good at doing simpler things before attempting at doing a "Frankenstein".

**Comment by heiwiper** [comment]
I would like to work on this feature request, there are two ideas that I get from this thread that seem to be related to incremental tree-style hinting:

1. Using the hint overlays only and discarding the ones that don't match what the used has clicked until the right hint overlay is selected. This is implemented in most hinting mechanisms like [qutebrowser](https://qutebrowser.org/) or [Link Hints](https://lydell.github.io/LinkHints/) from Firefox. As far as I know this isn't the default behaviour of nyxt, the user is expected to type the letters of the actual hint in the prompt buffer and then press `RET`.
2. Searching for substring through hintable elements and only display the ones. I've only seen this mechanism used in [Avy](https://github.com/abo-abo/avy) but I'm not sure if it would be convenient to handle hintable elements this way, this method would be more useful when moving a cursor to a specific text within the buffer but I might be wrong.

Please, let me know which method you would like to be implemented.

**Comment by jmercouris** [comment]
I'm not sure honestly, I use Avy personally. I think it is a better model.

**Comment by aadcg** [comment]
@heiwiper thank you for your interest! I think I understood exactly what you're asking, and both of your observations are spot on!

Indeed, a mechanism resembling avy wouldn't make much sense in this context.  

Option 1 would be an amazing improvement for our `hint-mode`. Notice that such a feature only makes sense when both `auto-follow-hints-p` and `fit-to-prompt-p` are both `t`. See `prompter:filter` in `source/mode/hint.lisp`:

```lisp
(prompter:filter
    (if (and (auto-follow-hints-p (find-submode 'hint-mode))
             (fit-to-prompt-p (find-submode 'hint-mode)))
        (lambda (suggestion source input)
          (declare (ignore source))
          (str:starts-with-p input
                             (prompter:attributes-default suggestion)
                             :ignore-case t))
        #'prompter:fuzzy-match))
```

**Comment by aadcg** [comment]
Closed via #2881.
